lib/bridge.js

```diff
@@ -773,61 +776,92 @@ Bridge.prototype._onEvent = function(event) {
+    const contextReady = this._getBridgeContext(event);
+    const dataReady = contextReady.then(c => ({
         request: request,
-        context: context
-    };
+        context: c
+    }));
 
-    var promise;
-    if (this.opts.disableContext) {
-        promise = Promise.resolve();
-        data.context = null;
-    }
-    else {
-        promise = context.get(this._roomStore, this._userStore);
-    }
+    const dataReadyLimited = this._limited(dataReady, request);
 
-    if (this.opts.queue.type === "none") { // consume as soon as we have context
-        promise.done(() => {
-            this._onConsume(null, data);
-        }, (err) => {
-            this._onConsume(err);
-        });
-        return request.getPromise();
-    }
+    this._queue.push(event, dataReadyLimited);
+    this._queue.consume();
+
+    return (
+        request
+        .getPromise()
+        .catch(EventNotHandledError, e => this._handleEventError(event, e))
+    );
+};
 
+/**
+ * Returns a promise that is like `promise` but adheres to all necessary limits.
+ */
+Bridge.prototype._limited = function(promise, request) {
+    // queue.perRequest controls whether multiple request can be processed by
+    // the bridge at once.
     if (this.opts.queue.perRequest) {
-        promise = Promise.settle([
-            promise,
-            this._prevRequestPromise
-        ]);
+        const promiseLimited = this._prevRequestPromise.reflect().return(promise);
         this._prevRequestPromise = request.getPromise();
+        return promiseLimited;
     }
 
-    this._queue.push(event, data, promise);
-    this._queue.consume();
-    return request.getPromise();
-};
+    return promise;
+}
 
 Bridge.prototype._onConsume = function(err, data) {
-    if (!err) {
-        this.opts.controller.onEvent(data.request, data.context);
+    if (err) {
+        // The data for the event could not be retrieved.
+        // In this case we reject the request and the event will be delivered
+        // by the HS to the bridge again.
+        data.request.reject();
+        this.opts.controller.onLog("onEvent failure: " + err);
         return;
     }
-    if (!this.opts.controller.onLog) {
-        return;
+
+    this.opts.controller.onEvent(data.request, data.context);
+};
+
+Bridge.prototype._getBridgeContext = function(event) {
+    if (this.opts.disableContext) {
+        return Promise.resolve(null);
     }
-    this.opts.controller.onLog(
-        "onEvent failure: " + err
+
+    const context = new BridgeContext({
+        sender: event.user_id,
+        target: event.state_key,
+        room: event.room_id
+    });
+    const contextReady = context.get(this._roomStore, this._userStore)
+
+    return contextReady.return(context);
+}
+
+Bridge.prototype._handleEventError = function(event, error) {
+    if (!error instanceof EventNotHandledError) {
+        error = wrap(error, InternalError);
+    }
+    this._botIntent.signalBridgeError(
```

Comment:
Do we need to handle cases where the bot may not be in the room?

Reply:
Yes. As this is more a design question and currently discussed in context of the MSC I might want to delay it here.

Reply:
You can put in a TODO line here just to explain that it's in progress.